### PR TITLE
iOS SDK: Namespace local storage of auth info by client app ID

### PR DESCRIPTION
**Pull request by jsflax**
Note: This does not include the migration. If we wish to include the migration, please update the JIRA ticket.

### CODE REVIEW

File: StitchCore/StitchCore/Source/Core/Storage.swift [hunk 1]
```diff
@@ -0,0 +1,58 @@
+import Foundation
+
+/// Storage allows us to store arbitrary data via
+/// an arbitrary source. The protocol is modeled
+/// after \(Foundation.UserDefaults).
+public protocol Storage {
```
Comment (edaniels) [inline StitchCore/StitchCore/Source/Core/Storage.swift]: Yay, not async!

**Comment by edaniels** [comment]
Note: We will be including an auto migration

**Comment by jsflax** [comment]
@edaniels @adamchel I've altered the original PR enough that I think this is worth a second look. PTAL

File: StitchCore/StitchCore/Source/Core/StitchClient.swift [hunk 2]
```diff
@@ -168,11 +167,32 @@ public class StitchClient: StitchClientType {
...
         self.appId = appId
         self.baseUrl = baseUrl
         self.networkAdapter = networkAdapter
+
+        let suiteName = "\(Consts.UserDefaultsName).\(appId)"
+        if let storage = storage {
+            self.storage = storage
+        } else {
+            #if !os(Linux)
+            guard let userDefaults = UserDefaults.init(suiteName: suiteName) else {
```
Comment (adamchel) [inline StitchCore/StitchCore/Source/Core/StitchClient.swift]: [style q] I saw this in several places, is there any reason we use the explicit `.init()` call rather than just calling the init function like `UserDefaults(suiteName: blah)`?

File: StitchCore/StitchCoreTests/AuthTests.swift [hunk 3]
```diff
@@ -124,16 +124,19 @@ class AuthTests: XCTestCase {
...
         let exp = expectation(description: "multiple logins work as expected")
-        let mlsStitchClient = StitchClient(appId: "stitch-tests-ios-sdk-jjmum")
         var anonUserId = ""
         var emailUserId = ""
-        
-        // check storage
-        XCTAssertFalse(mlsStitchClient.isAuthenticated)
-        XCTAssertNil(mlsStitchClient.loggedInProviderType)
-        
+        var mlsStitchClient: StitchClient!
```
Comment (adamchel) [inline StitchCore/StitchCoreTests/AuthTests.swift]: Not sure why these changes are showing up here. This was stuff from my multiple login semantics commit that made it conform to the factory changes. Can we make sure we rebase everything properly before merging?

File: StitchCore/StitchCoreTests/StitchCoreTests.swift [hunk 4]
```diff
@@ -25,6 +25,48 @@ class StitchCoreTests: XCTestCase {
     }
 
 
+    func testMultipleClientStorage() throws {
+        let exp = expectation(description: "multiple clients should store separately")
+
+        let creators = [StitchClientFactory.create(appId: "test1"),
+                        StitchClientFactory.create(appId: "test2")]
+
+        when(resolved: creators).done { (results: [Result<StitchClient>]) in
```
Comment (adamchel) [inline StitchCore/StitchCoreTests/StitchCoreTests.swift]: fancy 😮

File: StitchCore/StitchCoreTests/StitchCoreTests.swift [hunk 5]
```diff
@@ -25,6 +25,48 @@ class StitchCoreTests: XCTestCase {
...
+            XCTAssertEqual(sc1.storage.value(forKey: "fizz") as? String, "foo")
+            XCTAssertEqual(sc2.storage.value(forKey: "fizz") as? String, "bar")
+            exp.fulfill()
+        }
+
+        wait(for: [exp], timeout: 10)
+    }
+
+    func testMigrationLogic() throws {
+        let exp = expectation(description: "multiple clients should store separately")
```
Comment (adamchel) [inline StitchCore/StitchCoreTests/StitchCoreTests.swift]: [nit] the expectation description is the same as the previous test

File: StitchCore/StitchCore/Source/Core/StitchClient.swift (hunk 2 above)
Comment (edaniels) [reply StitchCore/StitchCore/Source/Core/StitchClient.swift]: I believe there's actually a difference where not specifying `init` causes it do be a coercion that calls init, not directly an init. I'm sure it's compiled to exactly the same thing though.

Comment (jsflax) [reply StitchCore/StitchCore/Source/Core/StitchClient.swift]: I believe @edaniels is correct. Personally, I do it because I find it to be more legible, and speaks closer to Swift's idioms– verbosity and readability over shortcuts.

File: StitchCore/StitchCore/Source/Core/StitchClient.swift [hunk 6]
```diff
@@ -168,11 +167,32 @@ public class StitchClient: StitchClientType {
...
+                self.storage = MemoryStorage.init(suiteName: suiteName)!
+                printLog(.warning, text: "Invalid suiteName: \(suiteName)")
+                printLog(.warning, text: "Defaulting to memory storage. NOTE: App will not persist data in this state")
```
[the remark this answers is not in the document]
Comment (jsflax) [reply StitchCore/StitchCore/Source/Core/StitchClient.swift]: Fair point. Will change the wording.

File: StitchCore/StitchCore/Source/Core/Storage.swift [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+}
+
+/// Failover implementation of the Storage protocol
+/// in the event that no other storage unit can
+/// be found.
+internal struct MemoryStorage: Storage {
+    /// Namespace of this suite
+    private let suiteName: String
+    /// Internal storage unit
+    fileprivate var storage: [String: Any] = [:]
```
[the remark this answers is not in the document]
Comment (jsflax) [reply StitchCore/StitchCore/Source/Core/Storage.swift]: 👍

File: StitchCore/StitchCoreTests/AuthTests.swift (hunk 3 above)
Comment (jsflax) [reply StitchCore/StitchCoreTests/AuthTests.swift]: Odd. Alright.

File: StitchCore/StitchCoreTests/StitchCoreTests.swift (hunk 5 above)
Comment (jsflax) [reply StitchCore/StitchCoreTests/StitchCoreTests.swift]: Good catch.

File: StitchCore/StitchCore/Source/Core/Storage.swift [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+        storage[versionKey] = latestVersion
+    #endif
+    // in future versions, `case 1:`, `case 2:` and so on
+    // could be added to perform similar migrations
+    default: break
+    }
+}
+
+internal struct StorageKeys {
+    internal let isLoggedInUDKey: String
```
Comment (edaniels) [inline StitchCore/StitchCore/Source/Core/Storage.swift]: [q] Given the struct is internal, does that not make the internal props internal too? Also, isn't internal the default access anyway?

File: StitchCore/StitchCore/Source/Core/StitchClient.swift [hunk 9]
```diff
@@ -5,41 +5,36 @@ import StitchLogger
 import Security
 import PromiseKit
 
-public struct Consts {
-    public static let DefaultBaseUrl =   "https://stitch.mongodb.com"
-    static let ApiPath =                 "/api/client/v2.0/"
+internal struct Consts {
```
[the remark this answers is not in the document]
Comment (jsflax) [reply StitchCore/StitchCore/Source/Core/StitchClient.swift]: I believe the original writer did this as a means of namespacing, yes. I've been slowly stripping these constants out, as most of them need to be used dynamically.

I would personally agree and say `internal let`s would be more appropriate here.

File: StitchCore/StitchCore/Source/Core/Storage.swift [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+    case .none:
+    #if !os(Linux)
+        // map each of the store's keys to a Promise
+        // that fetches the each value for each key,
+        // sets the old value to the new "namespaced" key
+        // remove the old key value pair,
+        // and set the version number
+        // we only care about UserDefaults here since no other
+        // `Storage` type was available before version 1
+        guard var originalStorage = UserDefaults.init(suiteName: "com.mongodb.stitch.sdk.UserDefaults") else {
```
[the remark this answers is not in the document]
Comment (jsflax) [reply StitchCore/StitchCore/Source/Core/Storage.swift]: Technically, it will always "run". But once the value is set, next time this block is hit, the version `case` will be the `default` case, which is empty. If it would be more legible, we could add a simple boolean check to see if `storage.value(forKey: versionKey)` is equal to `latestVersion`.

[the remark this answers is not in the document]
Comment (edaniels) [reply StitchCore/StitchCore/Source/Core/Storage.swift]: oh I missed that. fine as is!

File: StitchCore/StitchCore/Source/Core/StitchClient.swift [hunk 11]
```diff
@@ -5,41 +5,36 @@ import StitchLogger
...
+                       storage: Storage?) -> Promise<TClient>
 }
 
 public final class StitchClientFactory: StitchClientFactoryProtocol {
     typealias TClient = StitchClient
 
     static func create(appId: String,
-                       baseUrl: String = Consts.DefaultBaseUrl,
-                       networkAdapter: NetworkAdapter = StitchNetworkAdapter()) -> Promise<StitchClient> {
+                       baseUrl: String = "https://stitch.mongodb.com",
```
[the remark this answers is not in the document]
Comment (jsflax) [reply StitchCore/StitchCore/Source/Core/StitchClient.swift]: 👍

File: StitchCore/StitchCore/Source/Core/Storage.swift (hunk 8 above)
Comment (jsflax) [reply StitchCore/StitchCore/Source/Core/Storage.swift]: This ends up being a style choice.

File: StitchCore/StitchCore/Source/Core/StitchClient.swift (hunk 9 above)
[the remark this answers is not in the document]
Comment (jsflax) [reply StitchCore/StitchCore/Source/Core/StitchClient.swift]: I rescind the previous remark, as some of these do need to be exposed via default parameters.

In Swift (I'm not sure why this is the case), if you are using a default parameter, it must be of the same level of access (or higher) than the referenced method.

File: StitchCore/StitchCore/Source/Core/StitchHTTPClient.swift [hunk 12]
```diff
@@ -99,22 +108,23 @@ internal class StitchHTTPClient {
     internal func clearAuth() throws {
         authInfo = nil
 
-        try deleteToken(withKey: Consts.AuthRefreshTokenKey)
-        try deleteToken(withKey: Consts.IsLoggedInUDKey)
-        try deleteToken(withKey: Consts.AuthJwtKey)
+        try deleteToken(withKey: self.storageKeys.authRefreshTokenKey)
+        try deleteToken(withKey: self.storageKeys.isLoggedInUDKey)
```
[the remark this answers is not in the document]
Comment (jsflax) [reply StitchCore/StitchCore/Source/Core/StitchHTTPClient.swift]: I think this should be a separate ticket, but I agree with you.

**Comment by jsflax** [comment]
@adamchel PTAL
